ssl/s3_srvr.c

```diff
@@ -3012,19 +3024,32 @@ int ssl3_get_cert_verify(SSL *s)
         }
     } else
 #endif
-    if (pkey->type == NID_id_GostR3410_2001) {
-        unsigned char signature[64];
-        int idx;
+    if (pkey->type == NID_id_GostR3410_2001 
+            || pkey->type == NID_id_GostR3410_2012_256
+            || pkey->type == NID_id_GostR3410_2012_512) {
+        unsigned char signature[128];
+        size_t sigsize = (pkey->type == NID_id_GostR3410_2012_512) ? 128 : 64;
+        size_t dgstsize = (pkey->type == NID_id_GostR3410_2012_512) ? 64 : 32;
+
+        /*
+         * We can have
+         *    - client cert 34.10-94/2001 with 32-bytes digest offset 0
+         *    - client cert 34.10-2012 256 with 32-bytes digest offset 32
+         *    - client cert 34.10-2012 512 with 64-bytes digest offset 64
+         */
+        size_t offset = (pkey->type == NID_id_GostR3410_2012_512) ? 64
+            : ((pkey->type == NID_id_GostR3410_2012_256) ? 32 : 0);
+        size_t idx;
         EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new(pkey, NULL);
         EVP_PKEY_verify_init(pctx);
-        if (len != 64) {
+        if (len != sigsize) {
             fprintf(stderr, "GOST signature length is %d", len);
```

Comment:
This needs to log an error and goto f_err with al set to SSL_AD_DECODE_ERROR. There's a potential buffer overrun here otherwise